phpBB/phpbb/textreparser/row_based_plugin.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+*
+* This file is part of the phpBB Forum Software package.
+*
+* @copyright (c) phpBB Limited <https://www.phpbb.com>
+* @license GNU General Public License, version 2 (GPL-2.0)
+*
+* For full copyright and license information, please see
+* the docs/CREDITS.txt file.
+*
+*/
+
+namespace phpbb\textreparser;
+
+abstract class row_based_plugin extends base
+{
+	/**
+	* @var \phpbb\db\driver\driver_interface
+	*/
+	protected $db;
+
+	/**
+	* Constructor
+	*
+	* @param \phpbb\db\driver\driver_interface $db Database connection
+	*/
+	public function __construct(\phpbb\db\driver\driver_interface $db)
+	{
+		$this->db = $db;
+	}
+
+	/**
+	* Return the name of the column that correspond to each field
+	*
+	* @return array
+	*/
+	abstract public function get_columns();
+
+	/**
+	* Return the name of the table used by this plugin
+	*
+	* @return string
+	*/
+	abstract public function get_table_name();
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function get_max_id()
+	{
+		$columns = $this->get_columns();
+
+		$sql = 'SELECT MAX(' . $columns['id'] . ' AS max_id FROM ' . $this->get_table_name();
```

Comment:
You only get the max ID once then you work backward from it, like you would only get the timestamp once. Check out the last paragraph of the original post at https://area51.phpbb.com/phpBB/viewtopic.php?f=126&t=47666